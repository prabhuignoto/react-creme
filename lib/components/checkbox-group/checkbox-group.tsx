import classNames from "classnames";
import { nanoid } from "nanoid";
import React, { useCallback, useEffect, useMemo } from "react";
import { CheckBox } from "../checkbox/checkbox";
import { CheckboxModel } from "../checkbox/checkbox-model";
import { useFirstRender } from "../common/effects/useFirstRender";
import "./checkbox-group.scss";

interface CheckboxGroupProps {
  options: CheckboxModel[];
  border?: boolean;
  disabled?: boolean;
  layout?: "horizontal" | "vertical";
  checkboxStyle?: "square" | "round";
  onChange?: (
    selected: { id?: string; name?: string; isChecked: boolean }[]
  ) => void;
  noUniqueIds?: boolean;
}

const CheckBoxGroup: React.FunctionComponent<CheckboxGroupProps> = ({
  options = [],
  disabled,
  border,
  layout = "vertical",
  checkboxStyle = "square",
  onChange,
  noUniqueIds = false,
}) => {
  const wrapperClass = useMemo(() => {
    return classNames("rc-checkbox-group", {
      "rc-checkbox-group-border": border,
      "rc-checkbox-group-vertical": layout === "vertical",
      "rc-checkbox-group-horizontal": layout === "horizontal",
      "rc-checkbox-group-disabled": disabled,
    });
  }, [layout, disabled]);

  const [items, setItems] = React.useState(
    options
      ? options.map((option) => ({
          ...(!noUniqueIds ? { id: nanoid() } : { id: option.id }),
          ...option,
          isChecked: option.isChecked || false,
        }))
      : []
  );

  const handleChange = useCallback(
    (id: string, name: string, selected: boolean) => {
      setItems((items) => {
        const _newItems = items.map((item) => {
          return {
            ...item,
            isChecked: item.id === id ? selected : !!item.isChecked,
          };
        });

        return _newItems;
      });
    },
    []
  );

  useEffect(() => {
    if (!isFirstRender.current) {
      onChange?.(
        items.map((item) => ({
          id: item.id,
          name: item.label,
          isChecked: item.isChecked,
        }))
      );
    }
  }, [JSON.stringify(items.map((i) => i.isChecked))]);

  const isFirstRender = useFirstRender();

  return (
    <div className={wrapperClass} role="group">
      {items.map((item) => {
        const { id, ...rest } = item;
        return (
          <CheckBox
            key={id}
            {...rest}
            disabled={disabled || item.disabled}
            onChange={handleChange}
            checkBoxStyle={checkboxStyle}
            noUniqueId={true}
            id={item.id}
          />
        );
      })}
    </div>
  );
};

export { CheckBoxGroup };
