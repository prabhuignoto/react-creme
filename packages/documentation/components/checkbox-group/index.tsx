import React, { lazy } from "react";
import DemoPageRenderer from "../../common/demo-page-renderer";

function index() {
  return (
    <DemoPageRenderer
      title="Checkbox Group"
      description="Checkbox group is a group of checkboxes that can be used to select multiple options."
      demoWidget={lazy(() => import("./widgets"))}
      callbacks={[
        {
          name: "onChange",
          description:
            "Callback function to be called when the checkbox group changes.",
          default: "",
          optional: "Yes",
          type: "function",
        },
      ]}
      properties={[
        {
          name: "options",
          description: "Options to be displayed in the checkbox group.",
          default: "[]",
          optional: "Yes",
          type: "Array<Option>",
        },
        {
          name: "noUniqueIds",
          description:
            "If true, the checkbox group will not generate unique ids for each checkbox.",
          default: "false",
          optional: "Yes",
          type: "boolean",
        },
        {
          name: "checkboxStyle",
          description:
            "checkbox render style. <em>square</em> or <em>round</em>",
          default: "square",
          optional: "Yes",
          type: "string",
        },
        {
          name: "disabled",
          description: "Disables the checkbox group",
          default: "false",
          optional: "Yes",
          type: "boolean",
        },
        {
          name: "layout",
          description: "layout of the checkbox group. vertical or horizontal",
          default: "vertical",
          optional: "Yes",
          type: "string",
        },
      ]}
      tabTitles={["examples", "properties"]}
    ></DemoPageRenderer>
  );
}

export default index;
