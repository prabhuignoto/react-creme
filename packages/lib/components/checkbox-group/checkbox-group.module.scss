@use '@design/core.scss';
@use '@design/theme.scss';

.checkbox_group_wrapper {
  &.checkbox_group_horizontal {
    @extend %left;
  }

  &.checkbox_group_vertical {
    @extend %col;
  }
}
