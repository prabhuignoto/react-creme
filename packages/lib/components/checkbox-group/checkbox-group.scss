@use '../../design/core.scss';
@use '../../design/theme.scss';

.rc-checkbox-group-wrapper {
  &.rc-checkbox-group-horizontal {
    @extend %left;
  }

  &.rc-checkbox-group-vertical {
    @extend %col;
  }
}
