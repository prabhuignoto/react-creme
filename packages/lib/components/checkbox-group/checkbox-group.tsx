import classNames from "classnames";
import { nanoid } from "nanoid";
import React, { useCallback, useEffect, useMemo } from "react";
import { CheckBox } from "../checkbox/checkbox";
import { CheckboxModel } from "../checkbox/checkbox-model";
import { useFirstRender } from "../common/effects/useFirstRender";
import "./checkbox-group.scss";

export interface CheckboxGroupProps {
  options: CheckboxModel[];
  border?: boolean;
  disabled?: boolean;
  layout?: "horizontal" | "vertical";
  checkboxStyle?: "square" | "round";
  onChange?: (
    selected: { id?: string; name?: string; isChecked: boolean }[]
  ) => void;
  noUniqueIds?: boolean;
  RTL?: boolean;
}

const CheckBoxGroup: React.FunctionComponent<CheckboxGroupProps> = ({
  options = [],
  disabled,
  border,
  layout = "vertical",
  checkboxStyle = "square",
  onChange,
  noUniqueIds = false,
  RTL = false,
}) => {
  const wrapperClass = useMemo(() => {
    return classNames("rc-checkbox-group", {
      "rc-checkbox-group-border": border,
      "rc-checkbox-group-disabled": disabled,
      "rc-checkbox-group-horizontal": layout === "horizontal",
      "rc-checkbox-group-vertical": layout === "vertical",
    });
  }, [layout, disabled]);

  const [items, setItems] = React.useState(
    options
      ? options.map((option) => ({
          ...(!noUniqueIds ? { id: nanoid() } : { id: option.id }),
          ...option,
          isChecked: option.isChecked || false,
        }))
      : []
  );

  const handleChange = useCallback(
    (id: string, name: string, selected: boolean) => {
      setItems((items) => {
        const _newItems = items.map((item) => {
          return {
            ...item,
            isChecked: item.id === id ? selected : !!item.isChecked,
          };
        });

        return _newItems;
      });
    },
    []
  );

  useEffect(() => {
    if (!isFirstRender.current) {
      onChange?.(
        items.map((item) => ({
          id: item.id,
          isChecked: item.isChecked,
          name: item.label,
        }))
      );
    }
  }, [JSON.stringify(items.map((i) => i.isChecked))]);

  const isFirstRender = useFirstRender();

  return (
    <div className={wrapperClass} role="group">
      {items.map((item) => {
        const { id, ...rest } = item;
        return (
          <CheckBox
            key={id}
            {...rest}
            disabled={disabled || item.disabled}
            onChange={handleChange}
            checkBoxStyle={checkboxStyle}
            noUniqueId={true}
            id={item.id}
            RTL={RTL}
          />
        );
      })}
    </div>
  );
};

export { CheckBoxGroup };
