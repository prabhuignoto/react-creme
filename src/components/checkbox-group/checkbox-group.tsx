import classNames from "classnames";
import { nanoid } from "nanoid";
import React, { useMemo } from "react";
import { CheckBox } from "../checkbox/checkbox";
import { CheckboxModel } from "../checkbox/checkbox-model";
import "./checkbox-group.scss";

interface CheckboxGroupProps {
  options: CheckboxModel[];
  border?: boolean;
  disabled?: boolean;
  layout?: "horizontal" | "vertical";
}

const CheckBoxGroup: React.FunctionComponent<CheckboxGroupProps> = ({
  options,
  disabled,
  border,
  layout = "horizontal",
}) => {
  const wrapperClass = useMemo(() => {
    return classNames("rc-checkbox-group", {
      "rc-checkbox-group-border": border,
      "rc-checkbox-group-vertical": layout === "vertical",
      "rc-checkbox-group-horizontal": layout === "horizontal",
      "rc-checkbox-group-disabled": disabled,
    });
  }, [layout, disabled]);

  const [items, setItems] = React.useState(
    options ? options.map((option) => ({ id: nanoid(), ...option })) : []
  );

  const handleChange = (id: string, name: string, selected: boolean) => {
    const newItems = items.map((item) => {
      if (item.id === id) {
        return { ...item, isChecked: selected };
      }
      return item;
    });
    setItems(newItems);
  };

  return (
    <div className={wrapperClass} role="group">
      {items.map((item) => {
        const { id, ...rest } = item;
        return (
          <CheckBox
            key={id}
            {...rest}
            disabled={disabled || item.disabled}
            onChange={handleChange}
          />
        );
      })}
    </div>
  );
};

export { CheckBoxGroup };
