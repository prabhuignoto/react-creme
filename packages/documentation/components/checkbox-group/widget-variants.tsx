import React from 'react';
import { CheckBoxGroup } from '../../../lib/components';

export const Default = (
  <CheckBoxGroup
    onChange={value => console.log(value)}
    options={[
      {
        label: 'Option 1',
      },
      {
        label: 'Option 2',
      },
      {
        label: 'Option 3',
      },
      {
        label: 'Option 4',
      },
    ]}
  />
);

export const PreSelected = (
  <CheckBoxGroup
    size="md"
    options={[
      {
        label: 'Option 1',
      },
      {
        label: 'Option 2',
      },
      {
        isChecked: true,
        label: 'This option is preselected',
      },
    ]}
  />
);

export const Disabled = (
  <CheckBoxGroup
    options={[
      {
        label: 'Option 1',
      },
      {
        disabled: true,
        label: 'This option is disabled',
      },
      {
        disabled: true,
        isChecked: true,
        label: 'This option is disabled but preselected',
      },
    ]}
  />
);

export const CustomStyle = (
  <CheckBoxGroup
    checkboxStyle="round"
    options={[
      {
        label: 'Option 1',
      },
      {
        label: 'Option 2',
      },
      {
        label: 'Option 3',
      },
      {
        label: 'Option 4',
      },
    ]}
  />
);

export const RTL = (
  <CheckBoxGroup
    RTL
    options={[
      {
        label: 'Option 1',
      },
      {
        label: 'Option 2',
      },
      {
        label: 'Option 3',
      },
      {
        label: 'Option 4',
      },
    ]}
  />
);
