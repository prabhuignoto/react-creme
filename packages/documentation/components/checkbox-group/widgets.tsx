import { Section } from '../../../lib/components';
import { DemoWidget } from '../../common/demo-widget';
import {
  CustomStyle,
  Default,
  Disabled,
  PreSelected,
  RTL,
} from './widget-variants';

function Widgets() {
  return (
    <div className="rc-demo-widgets">
      <Section title="Default Render">
        <DemoWidget name="CheckBoxGroup" width={100}>
          {Default}
        </DemoWidget>
      </Section>
      <Section title="Preselected Option">
        <DemoWidget name="CheckBoxGroup" width={200}>
          {PreSelected}
        </DemoWidget>
      </Section>
      <Section title="Disabled Option">
        <DemoWidget name="CheckBoxGroup" width={200}>
          {Disabled}
        </DemoWidget>
      </Section>
      <Section title="Checkbox Group - Round style">
        <DemoWidget name="CheckBoxGroup" width={150}>
          {CustomStyle}
        </DemoWidget>
      </Section>
      <Section title="Checkbox Group - RTL">
        <DemoWidget name="CheckBoxGroup" width={150}>
          {RTL}
        </DemoWidget>
      </Section>
    </div>
  );
}

export default Widgets;
