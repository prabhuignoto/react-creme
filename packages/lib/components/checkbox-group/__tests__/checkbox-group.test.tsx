import { act, fireEvent, render, waitFor } from '@testing-library/react';
import React from 'react';
import { vi } from 'vitest';
import { CheckBoxGroup } from '../checkbox-group';

const options = [
  {
    label: 'Option 1',
    id: '1',
  },
  {
    label: 'Option 2',
    isChecked: true,
    id: '2',
  },
  {
    label: 'Option 3',
    disabled: true,
    id: '3',
  },
];

describe('CheckboxGroup', () => {
  it('should render checkbox group', () => {
    const { getByRole, getByText, getAllByRole } = render(
      <CheckBoxGroup options={options} />
    );

    expect(getByText('Option 1')).toBeInTheDocument();
    expect(getByText('Option 2')).toBeInTheDocument();
    expect(getByText('Option 3')).toBeInTheDocument();

    expect(getByRole('group')).toBeInTheDocument();

    expect(getAllByRole('checkbox')).toHaveLength(3);
  });

  it('should match snapshot', () => {
    const { container } = render(<CheckBoxGroup options={options} />);

    expect(container).toMatchSnapshot();
  });

  it('should check for isChecked', () => {
    const { getAllByRole } = render(<CheckBoxGroup options={options} />);

    expect(getAllByRole('checkbox')[1]).toBeChecked();
  });

  it('check for disabled option', () => {
    const { getAllByRole } = render(<CheckBoxGroup options={options} />);

    expect(getAllByRole('checkbox')[2]).toHaveAttribute(
      'aria-disabled',
      'true'
    );
  });

  it('should handle on change', async () => {
    const optionsData = [
      {
        label: 'Option 1',
        id: '1',
      },
      {
        label: 'Option 2',
        isChecked: true,
        id: '2',
      },
      {
        label: 'Option 3',
        disabled: true,
        id: '3',
      },
    ];
    const handler = vi.fn();
    const { getAllByRole } = render(
      <CheckBoxGroup options={optionsData} onChange={handler} />
    );

    await waitFor(() => {
      expect(getAllByRole('checkbox')).toHaveLength(3);
    });

    act(() => {
      fireEvent.click(getAllByRole('checkbox')[0]);
    });

    await waitFor(() => {
      expect(handler).toHaveBeenCalled();
    });
  });
});
