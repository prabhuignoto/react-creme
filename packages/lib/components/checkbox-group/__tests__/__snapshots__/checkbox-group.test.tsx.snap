// Vitest Snapshot v1

exports[`CheckboxGroup > should match snapshot 1`] = `
<div>
  <div
    class="_checkbox_group_wrapper_15yvp_1 _checkbox_group_vertical_15yvp_1"
    role="group"
  >
    <div
      aria-checked="false"
      aria-labelledby="1"
      class="_wrapper_pk20g_49 _sm_pk20g_24 undefined _hover_pk20g_1"
      role="checkbox"
      style="outline: none; position: relative;"
      tabindex="0"
    >
      <div
        class="_checkbox_pk20g_49 _sm_pk20g_24"
      >
        <span
          aria-label="checkbox icon"
          class="_icon_pk20g_1 _sm_pk20g_24 undefined"
          role="img"
        >
          <svg
            class="prefix__feather prefix__feather-check"
            fill="transparent"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 6L9 17l-5-5"
            />
          </svg>
        </span>
      </div>
      <label
        class="_label_pk20g_12 _label_sm_pk20g_12"
        id="1"
      >
        Option 1
      </label>
      <span
        class="_focus_ring_4fy2t_1"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      aria-checked="true"
      aria-labelledby="2"
      class="_wrapper_pk20g_49 _sm_pk20g_24 undefined _hover_pk20g_1"
      role="checkbox"
      style="outline: none; position: relative;"
      tabindex="0"
    >
      <div
        class="_checkbox_pk20g_49 _sm_pk20g_24"
      >
        <span
          aria-label="checkbox icon"
          class="_icon_pk20g_1 _checked_pk20g_140 _sm_pk20g_24 undefined"
          role="img"
        >
          <svg
            class="prefix__feather prefix__feather-check"
            fill="transparent"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 6L9 17l-5-5"
            />
          </svg>
        </span>
      </div>
      <label
        class="_label_pk20g_12 _label_sm_pk20g_12"
        id="2"
      >
        Option 2
      </label>
      <span
        class="_focus_ring_4fy2t_1"
        style="width: 6px; height: 6px;"
      />
    </div>
    <div
      aria-checked="false"
      aria-disabled="true"
      aria-labelledby="3"
      class="_wrapper_pk20g_49 _sm_pk20g_24 _disabled_pk20g_58 undefined _hover_pk20g_1"
      role="checkbox"
    >
      <div
        class="_checkbox_pk20g_49 _sm_pk20g_24 _disabled_pk20g_58"
      >
        <span
          aria-label="checkbox icon"
          class="_icon_pk20g_1 _sm_pk20g_24 undefined"
          role="img"
        >
          <svg
            class="prefix__feather prefix__feather-check"
            fill="transparent"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M20 6L9 17l-5-5"
            />
          </svg>
        </span>
      </div>
      <label
        class="_label_pk20g_12 _label_sm_pk20g_12"
        id="3"
      >
        Option 3
      </label>
    </div>
  </div>
</div>
`;
